### PR TITLE
[GlobalOptimization] Add patterns for bubbling of transposes

### PR DESCRIPTION
**Pull request by qedawkins**

### CODE REVIEW

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: Good practice to add the reason for returning failure.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: I think you can just use `getDpsInit(resultIndex)` and avoid getting the `OpOperand`

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: Can you do this before you create the op instead of after?

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: Seems a bit convoluted. Going by the comments, the `init` is just the `init` of the original `transpose` op. Then you just adjust the indexing maps of all the operands of the producer generics and that should be your final `linalg.generic`. The logic here seems quite involved.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:284]: Hmmm, this logic is probably duplicated from MLIR for propagating collapse shapes past elementwise operations. Instead of reimplementing for specific case. maybe we can just expose the upstream method as a utility on `LinalgOp` and use that here?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Still reviewing, but leaving some comments.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:284]: The reason I reimplemented it this way is because I want to leave it as a `linalg.transpose` named op. The upstream method will create a `linalg.generic` op. I could try to add a callback to the upstream utility to allow for providing a callback to construct the reshaped op?

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: I'm not sure if I have cycles to review it, so I only point it out for now. This is not used, can we remove it?

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: oh oops, this was leftover from sdxl, good catch!

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:284]: There was a change upstream that modified the corresponding collapse method to create a named op if the original operation was a named op...

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: COmment is off I think

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:704]: A lot of this seems similar to the `DpsInput` case. Why is this separate? Can they be merged?

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: Are you making this generalization so that const-hoisting/const-eval can pick these up? Wouldnt it be simpler to just allow const-eval to pick up transposes?

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:987]: I was under the impression that the reshape propagation thing would be changed. Is that not pushed yet?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
@qedawkins not sure if you still have things in flight for this. Let me know when I should review this again.

**Comment by qedawkins** [comment]
> @qedawkins not sure if you still have things in flight for this. Let me know when I should review this again.

Yeah still addressing comments. I'll let you know when it's ready again.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: All of the patterns for propagation look for a `linalg.transpose` op. By generalizing it, none of the patterns will apply anymore, thus stopping propagation for that transpose.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:284]: Has the patch for it landed yet? Looking at the upstream pattern I still see this: https://github.com/llvm/llvm-project/blob/26101e8c50e07ea53869b0c7f343d76bfed61b02/mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp#L862

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:704]: This pattern is for `transpose(elementwise)`, where as the DpsInput pattern is for `elementwise(transpose)`, so I couldn't figure out how to structure the actual rewrite to do both. I went ahead and unified the unary elementwise check though.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:987]: yes, wasn't pushed yet. I dropped the collapsing patterns and only kept the expansion patterns.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: That only works if the generic has no reduction iterator types. If there are reduction iterators, we need to actually transpose the init of the generic (what this is doing). I didn't think it would be better to write a different pattern for elementwise ops, but I can do that if you think it would be better.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: Using the interchange helper also lets it update `linalg.index` ops too, and it updates the generic in-place, so I figured it would be less code duplication to do it like this. Because of the way the interchange is constructed, it shouldn't fail so I changed it to an assert and added some comments. If you think it's better to do it before I can still make this change though.

**Comment by qedawkins** [comment]
@MaheshRavishankar this is ready for another look whenever you get a chance. I left the comments that I just responded to open.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:284]: Not for the expansion stuff. Its only updated for the collapse variant  https://github.com/llvm/llvm-project/blob/26101e8c50e07ea53869b0c7f343d76bfed61b02/mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp#L1590

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:284]: I see, I think that logic only applies for cases like `linalg.add`/`linalg.sub` that don't need to update any indexing maps or fields. For `linalg.transpose` we need to update the `permutation` attribute. I can upstream this pattern as is though.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: Nit: This is not actually bubbling the transpose through. This is folding the transpose. I think this could just reuse the elementwise op fusion logic if you just generalize the transpose and fuse the operations.

I am going to assume all of this logic is correct, but would be good to not duplicate all the logic that should already be implemented upstream and reuse it (unless I am missing something).

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:571]: I think this too might be simpler if you just generalize the transpose and just call the upstream fusion transformation.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: Not sure why you need that though. You cant propagate this transpose through a constant or a global load anyway. So why is this even needed here?

**Review by MaheshRavishankar — COMMENTED** [review]
Ok, one more round of comments. Should be good to go after that. I think broadly this might be made much simpler (i.e. the addtions in this patch) but just reusing the upstream elementwise fusion patterns. Leaving comment. I am fine with doing it as a follow on, but will wait for response before stamp.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: I tried this but the upstream logic only works for elementwise ops and this pattern also works for producers with reduction loops. This kind of fusion is actually unique to transposes (and other data-movement only ops like broadcast) because it is only changing the order in which data is written back in, so I'm not sure there will be much to reuse.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks!

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:571]: I'm getting some confusing rewriter errors trying to do this generalization right now. I'm thinking it will be easier to change upstream to work off of linalg ops and then circle back and change this pattern.

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:571]: Added a TODO for it, marking this as resolved.

File: compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp]: I just went ahead and dropped the whole pattern. We can add it back later if needed.

**Pull request merged**
